### ISSUE TITLE
Filtering large numbers of facilities by polygonal area results in high latency

### ISSUE THREAD

**Issue by ColeKettler** [open-apparel-registry/open-apparel-registry#1080]
## Overview

When searching for facilities and filtering by polygonal area, queries run slowly enough that they begin to time out. Service and RDS instance CPU usage increases significantly and can cause containers to restart.

### Expected Behavior

Queries should not take long enough to time out or cause large amounts of hanging requests.

### Actual Behavior

Queries that filter by polygonal area have extremely high latency, up to and including timing out.

### Steps to Reproduce

* On staging, run a search that filters based on an area that returns a large number of matches. For example, this query should return over 2000 facilities:

https://staging.openapparel.org/?countries=BD&boundary=%7B%22type%22%3A%22Polygon%22%2C%22coordinates%22%3A%5B%5B%5B88.934326%2C25.055745%5D%2C%5B90.032959%2C22.329752%5D%2C%5B91.428223%2C22.380556%5D%2C%5B92.087402%2C23.211058%5D%2C%5B90.439453%2C24.986058%5D%2C%5B89.428711%2C25.473033%5D%2C%5B88.934326%2C25.055745%5D%5D%5D%7D

This should immediately cause significant latency on Staging. Zoom out 2 - 3 levels and you should begin to see 50x errors.

### Additional context

This results in symptoms similar to #1010. The common factor appears to be filtering by polygonal area in regions that contain a large number of facilities (e.g. Bangladesh).

Here's a a sample query that, when run >50 times, took 4 - 7 seconds to resolve and contributed to application timeouts. This query also filters by country, but the effect is reproducible even without the country filter:

```#sql
CREATE TEMP TABLE hex_grid (
    geom,
    mvt_geom
)
AS (
    SELECT
        geom,
        ST_AsMVTGeom (ST_Centroid (geom), ST_MakeEnvelope (10018754.171394622, 2504688.5428486555, 10331840.239250703, 2817774.6107047363))
    FROM
        generate_hexgrid (39135.758482010104,
            10018754.171394622,
            2504688.5428486555,
            10331840.239250703,
            2817774.6107047363)
);

CREATE INDEX hex_grid_idx ON hex_grid
USING gist (geom);

SELECT
    ST_AsMVT (q,
        'facilitygrid')
FROM (
    SELECT
        hex_grid.mvt_geom,
        count(location),
        ST_XMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as xmin,
        ST_YMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as ymin,
        ST_XMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as xmax,
        ST_YMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as ymax
    FROM
        hex_grid
        JOIN api_facility ON ST_Contains (ST_Transform (hex_grid.geom,
                4326),
            location)
    WHERE ("api_facility"."country_code" IN ('BD')
        AND ST_Within ("api_facility"."location", ST_GeomFromEWKB ('\x0103000020e6100000010000000500000005508c2c99ab5640d6c6d8092f353640cede196d55595640ed48f59d5fe4354032384a5e9d6b5640f94d61a582ce3840dba2cc0699d85640a1f5f065a2bc384005508c2c99ab5640d6c6d8092f353640'::bytea)))
    AND abs(ST_XMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) - ST_XMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326)))) < 180
GROUP BY
    hex_grid.mvt_geom,
    xmin,
    ymin,
    xmax,
    ymax) AS q;

```

**Comment by jwalgran** [comment]
Thank you for this valuable analysis.

TL;DR I thought indexing on country would be a quick fix. It doesn't appear to be.

My attention was drawn to the filter by country in addition to the filter by area. I checked and we don't yet have an index on this field.

```
openapparelregistry=> \d api_facility
...
Indexes:
    "api_facility_pkey" PRIMARY KEY, btree (id)
    "api_facility_created_from_id_key" UNIQUE CONSTRAINT, btree (created_from_id)
    "api_facility_id_cd5004e6_like" btree (id varchar_pattern_ops)
    "api_facility_location_id" gist (location)
    "api_facility_updated_at_3b3ba7c1" btree (updated_at)
...
```

I modified the query in the description and to EXPLAIN ANALYZE the select. Consistently getting about 2.2s in staging

```
CREATE TEMP TABLE hex_grid (
    geom,
    mvt_geom
)
AS (
    SELECT
        geom,
        ST_AsMVTGeom (ST_Centroid (geom), ST_MakeEnvelope (10018754.171394622, 2504688.5428486555, 10331840.239250703, 2817774.6107047363))
    FROM
        generate_hexgrid (39135.758482010104,
            10018754.171394622,
            2504688.5428486555,
            10331840.239250703,
            2817774.6107047363)
);

CREATE INDEX hex_grid_idx ON hex_grid
USING gist (geom);

EXPLAIN ANALYZE
SELECT
    ST_AsMVT (q,
        'facilitygrid')
FROM (
    SELECT
        hex_grid.mvt_geom,
        count(location),
        ST_XMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as xmin,
        ST_YMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as ymin,
        ST_XMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as xmax,
        ST_YMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as ymax
    FROM
        hex_grid
        JOIN api_facility ON ST_Contains (ST_Transform (hex_grid.geom,
                4326),
            location)
    WHERE ("api_facility"."country_code" IN ('BD')
        AND ST_Within ("api_facility"."location", ST_GeomFromEWKB ('\x0103000020e6100000010000000500000005508c2c99ab5640d6c6d8092f353640cede196d55595640ed48f59d5fe4354032384a5e9d6b5640f94d61a582ce3840dba2cc0699d85640a1f5f065a2bc384005508c2c99ab5640d6c6d8092f353640'::bytea)))
    AND abs(ST_XMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) - ST_XMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326)))) < 180
GROUP BY
    hex_grid.mvt_geom,
    xmin,
    ymin,
    xmax,
    ymax) AS q;


```
```                                                                                                                                                        QUERY PLAN

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=56.38..56.39 rows=1 width=32) (actual time=2238.590..2238.590 rows=1 loops=1)
   ->  Subquery Scan on q  (cost=56.21..56.38 rows=1 width=56) (actual time=2237.745..2238.541 rows=17 loops=1)
         ->  GroupAggregate  (cost=56.21..56.37 rows=1 width=72) (actual time=2237.737..2238.529 rows=17 loops=1)
               Group Key: hex_grid.mvt_geom, (st_xmin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_
xmax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d))
               ->  Sort  (cost=56.21..56.21 rows=1 width=96) (actual time=2237.731..2237.829 rows=2226 loops=1)
                     Sort Key: hex_grid.mvt_geom, (st_xmin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)),
 (st_xmax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d))
                     Sort Method: quicksort  Memory: 410kB
                     ->  Nested Loop  (cost=0.28..56.20 rows=1 width=96) (actual time=1.394..2235.778 rows=2226 loops=1)
                           Join Filter: ((st_transform(hex_grid.geom, 4326) ~ api_facility.location) AND _st_contains(st_transform(hex_grid.geom, 4326), api_facility.location))
                           Rows Removed by Join Filter: 309974
                           ->  Index Scan using api_facility_location_id on api_facility  (cost=0.28..27.94 rows=1 width=32) (actual time=0.038..3.780 rows=2230 loops=1)
                                 Index Cond: ('0103000020E6100000010000000500000005508C2C99AB5640D6C6D8092F353640CEDE196D55595640ED48F59D5FE4354032384A5E9D6B5640F94D61A582CE3840DBA2CC06
99D85640A1F5F065A2BC384005508C2C99AB5640D6C6D8092F353640'::geometry ~ location)
                                 Filter: (((country_code)::text = 'BD'::text) AND _st_contains('0103000020E6100000010000000500000005508C2C99AB5640D6C6D8092F353640CEDE196D55595640ED48F59
D5FE4354032384A5E9D6B5640F94D61A582CE3840DBA2CC0699D85640A1F5F065A2BC384005508C2C99AB5640D6C6D8092F353640'::geometry, location))
                                 Rows Removed by Filter: 22
                           ->  Seq Scan on hex_grid  (cost=0.00..15.55 rows=47 width=64) (actual time=0.007..0.705 rows=140 loops=2230)
                                 Filter: (abs((st_xmax((st_envelope(st_transform(geom, 4326)))::box3d) - st_xmin((st_envelope(st_transform(geom, 4326)))::box3d))) < '180'::double precis
ion)
 Planning time: 0.243 ms
 Execution time: 2238.670 ms
```

Here is an explain analyze query where I removed the country filter and I am seeing consistent runs of about 700ms in staging. Still heavy, but about 3X faster. 

```
CREATE TEMP TABLE hex_grid (
    geom,
    mvt_geom
)
AS (
    SELECT
        geom,
        ST_AsMVTGeom (ST_Centroid (geom), ST_MakeEnvelope (10018754.171394622, 2504688.5428486555, 10331840.239250703, 2817774.6107047363))
    FROM
        generate_hexgrid (39135.758482010104,
            10018754.171394622,
            2504688.5428486555,
            10331840.239250703,
            2817774.6107047363)
);

CREATE INDEX hex_grid_idx ON hex_grid
USING gist (geom);

EXPLAIN ANALYZE
SELECT
    ST_AsMVT (q,
        'facilitygrid')
FROM (
    SELECT
        hex_grid.mvt_geom,
        count(location),
        ST_XMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as xmin,
        ST_YMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as ymin,
        ST_XMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as xmax,
        ST_YMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as ymax
    FROM
        hex_grid
        JOIN api_facility ON ST_Contains (ST_Transform (hex_grid.geom,
                4326),
            location)
    WHERE (ST_Within ("api_facility"."location", ST_GeomFromEWKB ('\x0103000020e6100000010000000500000005508c2c99ab5640d6c6d8092f353640cede196d55595640ed48f59d5fe4354032384a5e9d6b5640f94d61a582ce3840dba2cc0699d85640a1f5f065a2bc384005508c2c99ab5640d6c6d8092f353640'::bytea)))
    AND abs(ST_XMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) - ST_XMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326)))) < 180
GROUP BY
    hex_grid.mvt_geom,
    xmin,
    ymin,
    xmax,
    ymax) AS q;

```
```                                                                                                                                                        QUERY PLAN

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=132.60..132.61 rows=1 width=32) (actual time=707.943..707.943 rows=1 loops=1)
   ->  Subquery Scan on q  (cost=132.43..132.60 rows=1 width=56) (actual time=707.116..707.888 rows=17 loops=1)
         ->  GroupAggregate  (cost=132.43..132.59 rows=1 width=72) (actual time=707.109..707.877 rows=17 loops=1)
               Group Key: hex_grid.mvt_geom, (st_xmin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_
xmax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d))
               ->  Sort  (cost=132.43..132.44 rows=1 width=96) (actual time=707.103..707.198 rows=2226 loops=1)
                     Sort Key: hex_grid.mvt_geom, (st_xmin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)),
 (st_xmax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d))
                     Sort Method: quicksort  Memory: 410kB
                     ->  Nested Loop  (cost=0.28..132.42 rows=1 width=96) (actual time=248.625..705.763 rows=2226 loops=1)
                           Join Filter: ((st_transform(hex_grid.geom, 4326) ~ api_facility.location) AND _st_contains(st_transform(hex_grid.geom, 4326), api_facility.location))
                           Rows Removed by Join Filter: 309974
                           ->  Seq Scan on hex_grid  (cost=0.00..15.55 rows=47 width=64) (actual time=0.492..1.457 rows=140 loops=1)
                                 Filter: (abs((st_xmax((st_envelope(st_transform(geom, 4326)))::box3d) - st_xmin((st_envelope(st_transform(geom, 4326)))::box3d))) < '180'::double precis
ion)
                           ->  Materialize  (cost=0.28..27.93 rows=7 width=32) (actual time=0.001..0.139 rows=2230 loops=140)
                                 ->  Index Scan using api_facility_location_id on api_facility  (cost=0.28..27.89 rows=7 width=32) (actual time=0.116..2.255 rows=2230 loops=1)
                                       Index Cond: ('0103000020E6100000010000000500000005508C2C99AB5640D6C6D8092F353640CEDE196D55595640ED48F59D5FE4354032384A5E9D6B5640F94D61A582CE3840DB
A2CC0699D85640A1F5F065A2BC384005508C2C99AB5640D6C6D8092F353640'::geometry ~ location)
                                       Filter: _st_contains('0103000020E6100000010000000500000005508C2C99AB5640D6C6D8092F353640CEDE196D55595640ED48F59D5FE4354032384A5E9D6B5640F94D61A582
CE3840DBA2CC0699D85640A1F5F065A2BC384005508C2C99AB5640D6C6D8092F353640'::geometry, location)
                                       Rows Removed by Filter: 22
 Planning time: 0.208 ms
 Execution time: 708.039 ms
```

I tried adding a test index

```
create index TEST_IDX_COUNTRY on api_facility (country_code);
```

But the run time of the original query is the same

```
CREATE TEMP TABLE hex_grid (
    geom,
    mvt_geom
)
AS (
    SELECT
        geom,
        ST_AsMVTGeom (ST_Centroid (geom), ST_MakeEnvelope (10018754.171394622, 2504688.5428486555, 10331840.239250703, 2817774.6107047363))
    FROM
        generate_hexgrid (39135.758482010104,
            10018754.171394622,
            2504688.5428486555,
            10331840.239250703,
            2817774.6107047363)
);

CREATE INDEX hex_grid_idx ON hex_grid
USING gist (geom);

EXPLAIN ANALYZE
SELECT
    ST_AsMVT (q,
        'facilitygrid')
FROM (
    SELECT
        hex_grid.mvt_geom,
        count(location),
        ST_XMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as xmin,
        ST_YMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as ymin,
        ST_XMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as xmax,
        ST_YMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) as ymax
    FROM
        hex_grid
        JOIN api_facility ON ST_Contains (ST_Transform (hex_grid.geom,
                4326),
            location)
    WHERE ("api_facility"."country_code" IN ('BD')
        AND ST_Within ("api_facility"."location", ST_GeomFromEWKB ('\x0103000020e6100000010000000500000005508c2c99ab5640d6c6d8092f353640cede196d55595640ed48f59d5fe4354032384a5e9d6b5640f94d61a582ce3840dba2cc0699d85640a1f5f065a2bc384005508c2c99ab5640d6c6d8092f353640'::bytea)))
    AND abs(ST_XMax (ST_Envelope (ST_Transform (hex_grid.geom, 4326))) - ST_XMin (ST_Envelope (ST_Transform (hex_grid.geom, 4326)))) < 180
GROUP BY
    hex_grid.mvt_geom,
    xmin,
    ymin,
    xmax,
    ymax) AS q;
```
```                                                                                                                                                        QUERY PLAN

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=56.38..56.39 rows=1 width=32) (actual time=2219.601..2219.601 rows=1 loops=1)
   ->  Subquery Scan on q  (cost=56.21..56.38 rows=1 width=56) (actual time=2218.752..2219.554 rows=17 loops=1)
         ->  GroupAggregate  (cost=56.21..56.37 rows=1 width=72) (actual time=2218.745..2219.543 rows=17 loops=1)
               Group Key: hex_grid.mvt_geom, (st_xmin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_
xmax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d))
               ->  Sort  (cost=56.21..56.21 rows=1 width=96) (actual time=2218.739..2218.846 rows=2226 loops=1)
                     Sort Key: hex_grid.mvt_geom, (st_xmin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymin((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)),
 (st_xmax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d)), (st_ymax((st_envelope(st_transform(hex_grid.geom, 4326)))::box3d))
                     Sort Method: quicksort  Memory: 410kB
                     ->  Nested Loop  (cost=0.28..56.20 rows=1 width=96) (actual time=2.116..2217.223 rows=2226 loops=1)
                           Join Filter: ((st_transform(hex_grid.geom, 4326) ~ api_facility.location) AND _st_contains(st_transform(hex_grid.geom, 4326), api_facility.location))
                           Rows Removed by Join Filter: 309974
                           ->  Index Scan using api_facility_location_id on api_facility  (cost=0.28..27.94 rows=1 width=32) (actual time=0.051..3.377 rows=2230 loops=1)
                                 Index Cond: ('0103000020E6100000010000000500000005508C2C99AB5640D6C6D8092F353640CEDE196D55595640ED48F59D5FE4354032384A5E9D6B5640F94D61A582CE3840DBA2CC06
99D85640A1F5F065A2BC384005508C2C99AB5640D6C6D8092F353640'::geometry ~ location)
                                 Filter: (((country_code)::text = 'BD'::text) AND _st_contains('0103000020E6100000010000000500000005508C2C99AB5640D6C6D8092F353640CEDE196D55595640ED48F59
D5FE4354032384A5E9D6B5640F94D61A582CE3840DBA2CC0699D85640A1F5F065A2BC384005508C2C99AB5640D6C6D8092F353640'::geometry, location))
                                 Rows Removed by Filter: 22
                           ->  Seq Scan on hex_grid  (cost=0.00..15.55 rows=47 width=64) (actual time=0.006..0.704 rows=140 loops=2230)
                                 Filter: (abs((st_xmax((st_envelope(st_transform(geom, 4326)))::box3d) - st_xmin((st_envelope(st_transform(geom, 4326)))::box3d))) < '180'::double precis
ion)
 Planning time: 0.527 ms
 Execution time: 2219.734 ms
```

**Comment by ColeKettler** [comment]
That's unfortunate that we can't solve this with a simple index. Thanks for the `EXPLAIN ANALYZE` queries as well, that sheds a lot of light on the problem. I'll run a few `EXPLAIN ANALYZE` cases and include their output in future issue descriptions.
